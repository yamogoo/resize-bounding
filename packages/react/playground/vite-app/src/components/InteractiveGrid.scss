.boundarize-app {
  display: flex;
  @include box(100%);

  * {
    user-select: none;
  }

  &--container {
    @include box(100%);
    padding: 20px;
    box-sizing: border-box;
  }

  &__layout {
    display: flex;
    flex-direction: column;
    @include box(100%);
    border-radius: 24px;
    overflow: hidden;
  }
}
