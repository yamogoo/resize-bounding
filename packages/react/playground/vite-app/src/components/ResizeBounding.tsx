import ResizeBoundingComponent, { type Props } from "react-resize-bounding";

const ResizeBounding = (props: Props) => {
  return (
    <ResizeBoundingComponent
      {...props}
      options={{
        knob: {
          show: true,
        },
      }}
    >
      {props.children}
    </ResizeBoundingComponent>
  );
};

export default ResizeBounding;
