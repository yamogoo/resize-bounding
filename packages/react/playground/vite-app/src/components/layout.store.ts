enum LayoutKeys {
  A = "a",
  B = "b",
  C = "c",
  D = "d",
  E = "e",
  F = "f",
}

enum ActionTypes {
  SET_A_WIDTH = "A_WIDTH",
  SET_A_HEIGHT = "A_HEIGHT",
  SET_B_WIDTH = "B_WIDTH",
  SET_B_HEIGHT = "B_HEIGHT",
  SET_C_WIDTH = "C_WIDTH",
  SET_C_HEIGHT = "C_HEIGHT",
  SET_D_WIDTH = "D_WIDTH",
  SET_D_HEIGHT = "D_HEIGHT",
  SET_E_WIDTH = "E_WIDTH",
  SET_E_HEIGHT = "E_HEIGHT",
  SET_F_WIDTH = "F_WIDTH",
  SET_F_HEIGHT = "F_HEIGHT",
}

type ContainerSizeKeys =
  | "width"
  | "minWidth"
  | "maxWidth"
  | "height"
  | "minHeight"
  | "maxHeight";
type ContainerSize = Record<ContainerSizeKeys, number>;

interface Layout
  extends Record<LayoutKeys, Partial<ContainerSize & { title: string }>> {}

interface Actions {
  type: ActionTypes;
  payload: number;
}

const initState: Layout = {
  a: {
    height: 200,
    minHeight: 200,
    maxHeight: 320,
    title: "A",
  },
  b: {
    width: 320,
    minWidth: 240,
    maxWidth: 640,
    title: "B",
  },
  c: {
    width: 360,
    title: "C",
  },
  d: {
    width: 320,
    minWidth: 240,
    maxWidth: 640,
    title: "D",
  },
  e: {
    height: 320,
    minHeight: 240,
    maxHeight: 400,
    width: 640,
    minWidth: 240,
    maxWidth: 640,
    title: "E",
  },
  f: {
    title: "F",
  },
};

export {
  LayoutKeys,
  ActionTypes,
  type ContainerSizeKeys,
  type Actions,
  type ContainerSize,
  type Layout,
  initState,
};
