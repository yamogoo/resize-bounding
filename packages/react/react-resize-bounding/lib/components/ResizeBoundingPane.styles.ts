import { CSSProperties } from "react";
import { PaneDirections, type SplitterPosition } from "../shared/typings";

export const paneBaseStyles = (
  size: number,
  areaWidth: number,
  position: SplitterPosition,
): Record<PaneDirections, CSSProperties> => {
  let _offset: string = "0px";

  switch (position) {
    case "central":
      _offset = `-${areaWidth / 2}px`;
      break;
    case "external":
      _offset = `-${(areaWidth + size) / 2}px`;
      break;
    case "internal":
      _offset = `-${(areaWidth - size) / 2}px`;
      break;
  }

  return {
    l: { top: "0px", left: _offset, width: `${areaWidth}px`, height: "100%" },
    r: { top: "0px", right: _offset, width: `${areaWidth}px`, height: "100%" },
    t: { left: "0px", top: _offset, width: "100%", height: `${areaWidth}px` },
    b: {
      left: "0px",
      bottom: _offset,
      width: "100%",
      height: `${areaWidth}px`,
    },
  };
};

export const splitterBaseStyles = (
  size: number,
  areaWidth: number,
): Record<PaneDirections, CSSProperties> => {
  const _size = `${size}px`;

  const _offset: string = `${(areaWidth - size) / 2}px`;

  return {
    l: { right: _offset, width: _size, height: "100%" },
    r: { left: _offset, width: _size, height: "100%" },
    t: { bottom: _offset, height: _size, width: "100%" },
    b: { top: _offset, height: _size, width: "100%" },
  };
};
