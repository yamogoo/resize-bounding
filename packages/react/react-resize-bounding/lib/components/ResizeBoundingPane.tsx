import {
  useMemo,
  useRef,
  useState,
  MutableRefObject,
  type PropsWithChildren,
  type CSSProperties,
  useEffect,
} from "react";

import {
  PaneDirections,
  PaneDirectionAliases,
  type PaneDirectionKey,
  type IResizeBoundingClassNames,
  type IStyles,
  type Options,
} from "../shared/typings";

import {
  paneBaseStyles,
  splitterBaseStyles,
} from "./ResizeBoundingPane.styles";

export interface PaneEmittedData {
  x: number;
  y: number;
  dir: string;
}

export interface Props extends PropsWithChildren {
  prefix: string;
  direction: PaneDirectionKey;
  options: Options;
  classNames: IResizeBoundingClassNames;
  styles?: Partial<IStyles>;
  dragStart: (data: PaneEmittedData) => void;
  dragMove: (data: PaneEmittedData) => void;
  dragEnd: (data: PaneEmittedData) => void;
  focus: (state: boolean) => void;
}

const checkIsHorizontal = (direction: string): boolean =>
  new RegExp(
    `[${PaneDirections.LEFT} | ${PaneDirections.RIGHT} | ${PaneDirectionAliases.HORIZONTAL}]`,
  ).test(direction);

const checkIsVertical = (direction: string): boolean =>
  new RegExp(
    `[${PaneDirections.TOP} | ${PaneDirections.BOTTOM} | ${PaneDirectionAliases.VERTICAL}]`,
  ).test(direction);

const ResizeBoundingPane = ({
  children,
  classNames,
  direction,
  options,
  dragStart,
  dragMove,
  dragEnd,
  focus,
}: Props) => {
  const refPane: MutableRefObject<HTMLDivElement | null> = useRef(null);

  const [isFocused, setIsFocused] = useState(false);
  const [isPressed, setIsPressed] = useState(false);

  let isResizing = false;

  const paneComputedStyle = useMemo(() => {
    const _width = options?.width ?? 1;

    if (refPane && _width) {
      const _areaWidth = options.activeAreaWidth ?? options.width;
      const _styles = paneBaseStyles(
        _width,
        _areaWidth,
        options?.position ?? "central",
      );
      const value: CSSProperties = _styles[direction as PaneDirections];
      return value;
    }

    return;
  }, [options.width, options.activeAreaWidth, options?.position, direction]);

  const splitterComputedStyle = useMemo(() => {
    const _width = options?.width;

    if (refPane && _width) {
      const _areaWidth = options.activeAreaWidth ?? options.width;

      const _styles = splitterBaseStyles(_width, _areaWidth);
      const value: CSSProperties = _styles[direction as PaneDirections];
      return value;
    }
  }, [options.width, options.activeAreaWidth, direction]);

  const containerComputedStyles = useMemo(() => {
    const isHorizontal = checkIsHorizontal(direction);
    return { transform: `rotate(${isHorizontal ? 90 : 0}deg)` };
  }, [direction]);

  const updateCursor = (state: boolean) => {
    let cursorActive: string;

    if (checkIsHorizontal(direction)) {
      cursorActive = options?.cursor?.horizontal ?? "auto";
    } else if (checkIsVertical(direction)) {
      cursorActive = options?.cursor?.vertical ?? "auto";
    } else cursorActive = "auto";

    if (refPane.current)
      refPane.current.style.cursor = state ? cursorActive : "auto";
  };

  /* * * Events * * */

  const emitFocus = (state: boolean) => {
    focus(state);
  };

  const onFocus = (e: PointerEvent, isFocused: boolean): void => {
    e.stopPropagation();
    onSelected(isFocused);
  };

  const onSelected = (state: boolean): void => {
    setIsFocused(state);
    updateCursor(state);
    emitFocus(state);
  };

  const onDragStart = (e: PointerEvent): void => {
    e.preventDefault();
    e.stopImmediatePropagation();

    if (!options.touchActions && e.pointerType === "touch") return;

    isResizing = true;
    setIsPressed(true);
    onSelected(isResizing);

    const el = e.currentTarget as HTMLDivElement;
    el.setPointerCapture(e.pointerId);

    dragStart({
      x: Math.round(e.clientX),
      y: Math.round(e.clientY),
      dir: direction,
    });

    const onDragMove = (e: PointerEvent): void => {
      dragMove({
        x: Math.round(e.clientX),
        y: Math.round(e.clientY),
        dir: direction,
      });
    };

    const onDragEnd = (e: PointerEvent): void => {
      isResizing = false;
      setIsPressed(false);

      const el = e.currentTarget as HTMLDivElement;
      el.releasePointerCapture(e.pointerId);

      el.removeEventListener("pointermove", onDragMove);
      el.removeEventListener("pointerup", onDragEnd);

      dragEnd({
        x: Math.round(e.clientX),
        y: Math.round(e.clientY),
        dir: direction,
      });
    };

    el.addEventListener("pointermove", onDragMove);
    el.addEventListener("pointerup", onDragEnd);
  };

  const onDragCancel = (e: PointerEvent): void => {
    const el = e.currentTarget as HTMLDivElement;

    isResizing = false;
    el.releasePointerCapture(e.pointerId);

    dragEnd({
      x: Math.round(e.clientX),
      y: Math.round(e.clientY),
      dir: direction,
    });
  };

  const addEventListeners = () => {
    const el = refPane.current;

    if (el) {
      el.addEventListener("pointerenter", (e) => onFocus(e, true));
      el.addEventListener("pointerleave", (e) => onFocus(e, false));

      el.addEventListener("pointerdown", onDragStart);
      el.addEventListener("pointercancel", onDragCancel);
    }
  };

  const removeEventListeners = () => {
    const el = refPane.current;

    if (el) {
      el.removeEventListener("pointerenter", (e) => onFocus(e, true));
      el.removeEventListener("pointerleave", (e) => onFocus(e, false));

      el.removeEventListener("pointerdown", onDragStart);
      el.addEventListener("pointercancel", onDragCancel);
    }
  };

  useEffect(() => {
    addEventListeners();

    return () => {
      removeEventListeners();
    };
  });

  return (
    <div
      ref={refPane}
      data-testid="resize-bounding-pane"
      className={`${classNames.pane}${isPressed || isFocused ? " active" : ""}${options.addStateClasses ? ` ${isPressed ? "pressed" : isFocused ? "focused" : "normal"}` : ""}`}
      style={paneComputedStyle}
    >
      <div
        data-testid="resize-bounding-splitter"
        className={classNames.splitter}
        style={splitterComputedStyle}
      >
        {(isFocused || !options?.knob?.normalHidden) && options?.knob?.show && (
          <div
            data-testid="resize-bounding-splitter-container"
            className={classNames.splitterContainer}
            style={containerComputedStyles}
          >
            <div data-testid="resize-bounding-knob" className={classNames.knob}>
              {children}
            </div>
          </div>
        )}
      </div>
    </div>
  );
};

export default ResizeBoundingPane;
