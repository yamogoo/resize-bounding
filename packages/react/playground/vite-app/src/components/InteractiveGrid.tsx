import { useReducer } from "react";

import ResizeBounding from "react-resize-bounding";
import deepmerge from "deepmerge";

import "./InteractiveGrid.scss";

import UIInnerBlock from "@/components/helpers/UIInnerBlock";
import UISizeField from "@/components/helpers/UISizeField";

import {
  type Actions,
  type ContainerSizeKeys,
  initState,
  LayoutKeys,
  ActionTypes,
} from "./layout.store";

const layoutReducer = (
  state: typeof initState,
  action: Actions,
): typeof initState => {
  const { type, payload } = action;
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const [sectionKey, propKey] = type.toLowerCase().split("_");

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  for (const [_, v] of Object.entries(LayoutKeys)) {
    if (sectionKey === v) {
      return deepmerge(
        state,
        (state[sectionKey][propKey as ContainerSizeKeys] = payload),
      );
    }
  }

  throw new Error("Invalid action name!");
};

const InteractiveGrid = () => {
  const [layout, dispatch] = useReducer(layoutReducer, initState);

  return (
    <main className="boundarize-app">
      <div className="boundarize-app--container">
        <div className="boundarize-app__layout">
          <ResizeBounding
            height={layout.a.height}
            minHeight={layout.a.minHeight}
            maxHeight={layout.a.maxHeight}
            directions="b"
            style={{ borderBottom: "1px solid gray" }}
            updateHeight={(height) => {
              dispatch({ type: ActionTypes.SET_A_HEIGHT, payload: height });
            }}
          >
            <UIInnerBlock title={layout.a.title}>
              <UISizeField
                height={layout.a.height}
                updateHeight={(height) => {
                  layout.a.height = height;
                }}
              />
            </UIInnerBlock>
          </ResizeBounding>
          <ResizeBounding
            directions="''"
            style={{ display: "flex", height: "100%" }}
          >
            <ResizeBounding
              width={layout.b.width}
              min-width={layout.b.minWidth}
              max-width={layout.b.maxWidth}
              directions="'r'"
              style={{ display: "flex", borderRight: "1px solid gray" }}
              updateWidth={(width) => {
                dispatch({ type: ActionTypes.SET_B_WIDTH, payload: width });
              }}
            >
              <UIInnerBlock title={layout.b.title}>
                <UISizeField
                  width={layout.b.width}
                  updateWidth={(width) => {
                    layout.b.width = width;
                  }}
                />
              </UIInnerBlock>
            </ResizeBounding>
            <UIInnerBlock title={layout.c.title}>
              <UISizeField />
            </UIInnerBlock>
            <ResizeBounding
              width={layout.d.width}
              min-width={layout.d.minWidth}
              max-width={layout.d.maxWidth}
              directions="'l'"
              style={{ borderLeft: "1px solid gray" }}
              updateWidth={(width) => {
                dispatch({ type: ActionTypes.SET_D_WIDTH, payload: width });
              }}
            >
              <UIInnerBlock title={layout.d.title}>
                <UISizeField
                  width={layout.d.width}
                  updateWidth={(width) => {
                    layout.d.width = width;
                  }}
                />
              </UIInnerBlock>
            </ResizeBounding>
          </ResizeBounding>
          <ResizeBounding
            height={layout.e.height}
            min-height={layout.e.minHeight}
            max-height={layout.e.maxHeight}
            directions="'t'"
            style={{
              display: "flex",
              width: "100%",
              borderTop: "1px solid gray",
            }}
            updateHeight={(height) => {
              dispatch({ type: ActionTypes.SET_E_HEIGHT, payload: height });
            }}
          >
            <ResizeBounding
              width={layout.e.width}
              min-width={layout.e.minWidth}
              max-width={layout.e.maxWidth}
              directions="'r'"
              style={{
                borderRight: "1px solid gray",
              }}
              updateWidth={(width) => {
                dispatch({ type: ActionTypes.SET_E_WIDTH, payload: width });
              }}
            >
              <UIInnerBlock title={layout.e.title}>
                <UISizeField
                  width={layout.e.width}
                  height={layout.e.height}
                  updateWidth={(width) => {
                    layout.e.width = width;
                  }}
                  updateHeight={(height) => {
                    layout.e.height = height;
                  }}
                />
              </UIInnerBlock>
            </ResizeBounding>
            <ResizeBounding
              directions="''"
              style={{ display: "flex", width: "100%" }}
            >
              <UIInnerBlock title={layout.f.title}>
                <UISizeField
                  height={layout.e.height}
                  updateHeight={(height) => {
                    dispatch({
                      type: ActionTypes.SET_E_HEIGHT,
                      payload: height,
                    });
                  }}
                />
              </UIInnerBlock>
            </ResizeBounding>
          </ResizeBounding>
        </div>
      </div>
    </main>
  );
};

export default InteractiveGrid;
