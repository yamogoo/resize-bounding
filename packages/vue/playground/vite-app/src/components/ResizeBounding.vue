<template>
  <ResizeBounding
    :options="{
      knob: {
        show: true,
      },
    }"
  >
    <slot></slot>
  </ResizeBounding>
</template>

<script setup lang="ts">
import ResizeBounding from "vue3-resize-bounding";
</script>
