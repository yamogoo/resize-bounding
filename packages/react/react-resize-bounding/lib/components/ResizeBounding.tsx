import {
  CSSProperties,
  MutableRefObject,
  useMemo,
  useRef,
  type PropsWithChildren,
} from "react";
import deepmerge from "deepmerge";

import {
  PaneDirectionAliases,
  PaneDirections,
  type IStyles,
  type Options,
} from "../shared/typings";
import {
  defaultOptions,
  getClassNames,
  PREFIX,
} from "./ResizeBounding.classNames";

import ResizeBoundingPane, {
  PaneEmittedData,
  type PaneDirectionKey,
} from "./ResizeBoundingPane";

interface Props extends PropsWithChildren {
  width: number;
  height: number;
  minWidth: number;
  maxWidth: number;
  minHeight: number;
  maxHeight: number;
  disabled: boolean;
  directions: PaneDirectionKey;
  alwaysShowKnob: boolean;
  options: Partial<Options>;
  style: CSSProperties;
  styles: Partial<IStyles>;
  updateWidth: (width: number) => void;
  updateHeight: (height: number) => void;
  dragStart: (dir: string) => void;
  dragMove: (dir: string) => void;
  dragEnd: (dir: string) => void;
  focus: (data: { state: boolean; direction: string }) => void;
}

const ResizeBounding = (props: Partial<Props>) => {
  const {
    width,
    minWidth = 0,
    maxWidth = Number.POSITIVE_INFINITY,
    height,
    minHeight = 0,
    maxHeight = Number.POSITIVE_INFINITY,
    styles,
    children,
    style,
    updateWidth,
    updateHeight,
    dragStart,
    dragMove,
    dragEnd,
    focus,
  } = props;

  let { width: newWidth, height: newHeight } = props;

  const refRoot: MutableRefObject<HTMLDivElement | null> = useRef(null);

  let prevWidth = newWidth,
    prevHeight = newHeight;

  const options = useMemo(() => {
    return deepmerge(defaultOptions, props.options ?? {});
  }, [props.options]);

  const classNames = getClassNames(styles ?? {}, options.prefix ?? PREFIX);

  const computedStyle = useMemo(() => {
    const _width = {
      width: `${width}px`,
      minWidth: `${width}px`,
    };

    const _height = {
      height: `${height}px`,
      minHeight: `${height}px`,
    };

    return {
      ...(width && _width),
      ...(height && _height),
    };
  }, [width, height]);

  const panes = useMemo(() => {
    const directions = props.directions ?? "";
    return {
      left: {
        show:
          RegExp(PaneDirections.LEFT).test(directions) ||
          RegExp(PaneDirectionAliases.HORIZONTAL).test(directions),
        direction: PaneDirections.LEFT,
      },
      right: {
        show:
          RegExp(PaneDirections.RIGHT).test(directions) ||
          RegExp(PaneDirectionAliases.HORIZONTAL).test(directions),
        direction: PaneDirections.RIGHT,
      },
      bottom: {
        show:
          RegExp(PaneDirections.BOTTOM).test(directions) ||
          RegExp(PaneDirectionAliases.VERTICAL).test(directions),
        direction: PaneDirections.BOTTOM,
      },
      top: {
        show:
          RegExp(PaneDirections.TOP).test(directions) ||
          RegExp(PaneDirectionAliases.VERTICAL).test(directions),
        direction: PaneDirections.TOP,
      },
    };
  }, [props.directions]);

  /* * * Events * * */

  let startWidth = props.width ?? 0,
    startHeight = props.height ?? 0;

  let startX = 0,
    startY = 0;

  const onDragStart = ({ x, y, dir }: PaneEmittedData): void => {
    startWidth = props.width ?? 0;
    startHeight = props.height ?? 0;

    startX = x;
    startY = y;

    typeof dragStart === "function" && dragStart(dir);
  };

  const truncateInRange = (
    min: number,
    max: number | undefined,
    next: number,
  ): number => {
    const _max = max ?? Number.POSITIVE_INFINITY;

    if (next <= min) return min;
    else if (next >= _max) return _max;

    return next;
  };

  const onDragMove = ({ x, y, dir }: PaneEmittedData): void => {
    if (!refRoot.current) return;
    typeof dragMove === "function" && dragMove(dir);

    if (dir === PaneDirections.LEFT) {
      newWidth = startWidth + (startX - x);

      if (prevWidth === newWidth) return;

      const truncated = truncateInRange(minWidth, maxWidth, newWidth);

      typeof updateWidth === "function" && updateWidth(truncated);
      prevWidth = truncated;
    } else if (dir === PaneDirections.RIGHT) {
      newWidth = startWidth + (x - startX);

      if (prevWidth === newWidth) return;

      const truncated = truncateInRange(minWidth, maxWidth, newWidth);

      typeof updateWidth === "function" && updateWidth(truncated);
      prevWidth = truncated;
    } else if (dir === PaneDirections.TOP) {
      newHeight = startHeight + (startY - y);

      if (prevHeight === newHeight) return;

      const truncated = truncateInRange(minHeight, maxHeight, newHeight);

      typeof updateHeight === "function" && updateHeight(truncated);
      prevHeight = truncated;
    } else if (dir === PaneDirections.BOTTOM) {
      newHeight = startHeight + (y - startY);

      if (prevHeight === newHeight) return;

      const truncated = truncateInRange(minHeight, maxHeight, newHeight);

      typeof updateHeight === "function" && updateHeight(truncated);
      prevHeight = truncated;
    }
  };

  return (
    <div
      ref={refRoot}
      data-testid="resize-bounding-container"
      className={classNames.container}
      style={{ ...style, ...computedStyle }}
    >
      {children}
      {Object.values(panes).map(({ show, direction }) => {
        return (
          show && (
            <ResizeBoundingPane
              key={direction}
              prefix={options.prefix ?? ""}
              direction={direction}
              options={options}
              styles={styles}
              classNames={classNames}
              focus={(state) => {
                typeof focus === "function" && focus({ state, direction });
              }}
              dragStart={onDragStart}
              dragMove={onDragMove}
              dragEnd={({ dir }) =>
                typeof dragEnd === "function" && dragEnd(dir)
              }
            >
              {direction}
            </ResizeBoundingPane>
          )
        );
      })}
    </div>
  );
};

export default ResizeBounding;
