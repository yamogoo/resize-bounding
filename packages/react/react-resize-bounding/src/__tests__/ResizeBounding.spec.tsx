import { test, expect, describe, beforeEach, afterEach, vi } from "vitest";
import {
  cleanup,
  createEvent,
  fireEvent,
  render,
  screen,
} from "@testing-library/react";
import "@testing-library/jest-dom";

import React, { CSSProperties } from "react";

import { DataTestIds, PREFIX } from "./setup";
import { PaneDirectionAliases, PaneDirections } from "../../lib/shared/typings";

import ResizeBounding from "../../lib/components/ResizeBounding";
import { type PaneEmittedData } from "../../lib/components/ResizeBoundingPane";

afterEach(() => {
  cleanup();
});

const emits = {
  dragStart: vi.fn(),
  dragMove: vi.fn(),
  dragEnd: vi.fn(),
  focus: vi.fn(),
};

describe("ResizeBounding", () => {
  beforeEach(() => {
    Element.prototype.setPointerCapture = vi.fn();
    Element.prototype.releasePointerCapture = vi.fn();
  });

  afterEach(() => {
    cleanup();
  });

  describe("components", () => {
    test.each([
      ...Object.values(PaneDirections),
      ...Object.values(PaneDirectionAliases),
    ])(
      `Should not display any selected borders when set to "disabled" (--%s)`,
      async (direction) => {
        const { container } = render(
          <ResizeBounding directions={direction} disabled={true} />,
        );

        const paneEl = container.querySelectorAll(
          `[data-testid="${DataTestIds.PANE}"]`,
        );

        expect(paneEl.length).toBe(0);
        expect(paneEl.length).toMatchSnapshot();
      },
    );

    test.each(["rl", "ltrb", "bltr", "bl", "tr"])(
      "should render %s panes (panes computed value)",
      (directions) => {
        const { container } = render(
          <ResizeBounding directions={directions} />,
        );

        const panEls = container.querySelectorAll(
          `[data-testid="${DataTestIds.PANE}"]`,
        );
        expect(panEls.length).toBe(directions.length);
        expect(panEls.length).toMatchSnapshot();
      },
    );

    test.each(["sd", "wo", ""])(
      "should not render any pane element",
      (directions) => {
        const { container } = render(
          <ResizeBounding directions={directions} />,
        );

        const panEls = container.querySelectorAll(
          `[data-testid="${DataTestIds.PANE}"]`,
        );
        expect(panEls.length).toBe(0);
        expect(panEls.length).toMatchSnapshot();
      },
    );
  });

  describe("slots", () => {
    const SLOT_CONTENT = "Inner Container";

    test.each([<p>{SLOT_CONTENT}</p>])(
      `should render default slot "${SLOT_CONTENT}"`,
      (slot) => {
        render(<ResizeBounding directions="">{slot}</ResizeBounding>);

        const slotText = screen.getByTestId(DataTestIds.ROOT).textContent;
        expect(slotText).toContain(SLOT_CONTENT);
        expect(slotText).toMatchSnapshot();
      },
    );

    test.each([<p>knob</p>])("should render default slot", (slot) => {
      const { container } = render(
        <ResizeBounding
          directions="r"
          options={{
            knob: {
              show: true,
            },
          }}
          knob={slot}
        />,
      );

      const text = container.querySelector(
        `[data-testid="${DataTestIds.KNOB}"]`,
      )?.textContent;

      expect(text).toBe("knob");
      expect(text).toMatchSnapshot();
    });
  });

  describe("classes", () => {
    describe("prefix", () => {
      test.each(["my-prefix"])("should have custom prefix (%s)", (prefix) => {
        render(<ResizeBounding options={{ prefix }} />);

        const rootEl = screen.getByTestId(DataTestIds.ROOT);
        const classes = rootEl.classList;

        expect(classes.toString()).toContain(prefix);
        expect(classes).toMatchSnapshot();
      });

      test("should have default prefix (%s)", () => {
        render(<ResizeBounding />);

        const rootEl = screen.getByTestId(DataTestIds.ROOT);
        const classes = rootEl.classList;

        expect(classes.toString()).toContain(PREFIX);
        expect(classes).toMatchSnapshot();
      });
    });
  });

  describe("events", () => {
    describe("drag", () => {
      test.each(Object.values(PaneDirections))(
        `should emit the "dragStart -> dragMove -> dragEnd" event, which takes a direction argument "%s"`,
        async (direction) => {
          const props = { ...emits, directions: direction };

          const checkEvent = async (
            node: HTMLElement,
            eventName: string,
            eventData: { clientX: number; clientY: number },
            callback: (data: PaneEmittedData) => void,
          ) => {
            const ev = createEvent[eventName](node, eventData);
            fireEvent(node, ev);

            expect(callback).toHaveBeenCalledWith(direction);
          };

          render(<ResizeBounding {...props} />);
          const paneEl = screen.getByTestId(DataTestIds.SPLITTER);

          checkEvent(
            paneEl,
            "pointerDown",
            { clientX: 1, clientY: 1 },
            props.dragStart,
          );

          checkEvent(
            paneEl,
            "pointerMove",
            { clientX: 2, clientY: 2 },
            props.dragMove,
          );

          const callback = checkEvent(
            paneEl,
            "pointerUp",
            { clientX: 3, clientY: 3 },
            props.dragEnd,
          );

          expect(callback).toMatchSnapshot();
        },
      );

      test.each([
        [
          {
            alias: PaneDirectionAliases.HORIZONTAL,
            directions: [PaneDirections.RIGHT, PaneDirections.LEFT],
          },
        ],
        [
          {
            alias: PaneDirectionAliases.VERTICAL,
            directions: [PaneDirections.TOP, PaneDirections.BOTTOM],
          },
        ],
      ])(
        `check aliases: should emit "dragStart" with direction argument "%o"`,
        ({ alias, directions }) => {
          const props = { directions: alias, ...emits };
          render(<ResizeBounding {...props} />);

          directions.map((_d, idx) => {
            const paneEls = screen.getAllByTestId(DataTestIds.SPLITTER);
            const pos = { x: 1, y: 1 };

            fireEvent.pointerDown(paneEls[idx], {
              clientX: pos.x,
              clientY: pos.y,
            });

            expect(props.dragStart).toHaveBeenCalledWith(_d);
          });
        },
      );
    });
  });

  describe("styles", () => {
    describe("container", () => {
      test.each([{ background: "blue", display: "flex" }])(
        "should apply inline styles (%s)",
        (containerStyles: CSSProperties) => {
          render(<ResizeBounding style={containerStyles} />);

          const containerEl = screen.getByTestId(DataTestIds.ROOT);

          const styles = containerEl.getAttribute("style");

          expect(styles).toContain("display: flex");
          expect(styles).toContain("background: blue");
          expect(styles).toMatchSnapshot();
        },
      );
    });
  });
});
