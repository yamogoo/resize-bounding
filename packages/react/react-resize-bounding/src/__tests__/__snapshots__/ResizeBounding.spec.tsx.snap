// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ResizeBounding > classes > prefix > should have custom prefix (my-prefix) 1`] = `
DOMTokenList {
  "0": "my-prefixcontainer",
  "1": "my-prefixcontainer-5",
}
`;

exports[`ResizeBounding > classes > prefix > should have default prefix (%s) 1`] = `
DOMTokenList {
  "0": "resize-bounding-container",
  "1": "resize-bounding-container-0",
}
`;

exports[`ResizeBounding > components > Should not display any selected borders when set to "disabled" (--b) 1`] = `0`;

exports[`ResizeBounding > components > Should not display any selected borders when set to "disabled" (--h) 1`] = `0`;

exports[`ResizeBounding > components > Should not display any selected borders when set to "disabled" (--l) 1`] = `0`;

exports[`ResizeBounding > components > Should not display any selected borders when set to "disabled" (--r) 1`] = `0`;

exports[`ResizeBounding > components > Should not display any selected borders when set to "disabled" (--t) 1`] = `0`;

exports[`ResizeBounding > components > Should not display any selected borders when set to "disabled" (--v) 1`] = `0`;

exports[`ResizeBounding > components > should not render any pane element 1`] = `0`;

exports[`ResizeBounding > components > should not render any pane element 2`] = `0`;

exports[`ResizeBounding > components > should not render any pane element 3`] = `0`;

exports[`ResizeBounding > components > should render bl panes (panes computed value) 1`] = `2`;

exports[`ResizeBounding > components > should render bltr panes (panes computed value) 1`] = `4`;

exports[`ResizeBounding > components > should render ltrb panes (panes computed value) 1`] = `4`;

exports[`ResizeBounding > components > should render rl panes (panes computed value) 1`] = `2`;

exports[`ResizeBounding > components > should render tr panes (panes computed value) 1`] = `2`;

exports[`ResizeBounding > events > drag > should emit the "dragStart -> dragMove -> dragEnd" event, which takes a direction argument "b" 1`] = `Promise {}`;

exports[`ResizeBounding > events > drag > should emit the "dragStart -> dragMove -> dragEnd" event, which takes a direction argument "l" 1`] = `Promise {}`;

exports[`ResizeBounding > events > drag > should emit the "dragStart -> dragMove -> dragEnd" event, which takes a direction argument "r" 1`] = `Promise {}`;

exports[`ResizeBounding > events > drag > should emit the "dragStart -> dragMove -> dragEnd" event, which takes a direction argument "t" 1`] = `Promise {}`;

exports[`ResizeBounding > slots > should render default slot "Inner Container" 1`] = `"Inner Container"`;

exports[`ResizeBounding > slots > should render default slot 1`] = `"knob"`;

exports[`ResizeBounding > styles > container > should apply inline styles ({ background: 'blue', display: 'flex' }) 1`] = `"background: blue; display: flex;"`;
